### DISCUSSION TITLE
Getting real client IP in Nginx container for logging purposes

### DISCUSSION THREAD

**Discussion by bkouhen** [BretFisher/ama#173] [Q&A]
Hello Bret !

I have setup a web application behind an nginx proxy and both of these services are running through docker-compose (dev) or in a docker swarm (prod). I'm trying to add the ELK stack for logging and monitoring purposes but i can't get my hand on the real client IP (Public IP visiting the nginx server)

I've seen that this seems to be a recurrent issue in docker world because when i check the nginx service logs, i only see the container IP (for docker compose, like 172.18.0.1) or the overlay network IP (for docker swarm, like 10.0.0.2). I've tried many setups where i put my nginx service in host networking mode but i'm never able to see the real client IP and therefore i can't see which IPs are visiting my web app.

Would be great if you had some solution for that and maybe some explanation about how this works in detail

Thanks !

**Comment by BretFisher** [comment]
On a local machine, with Docker Desktop, that could be tricky, because the Linux VM is already behind NAT and there are mulltiple levels of NAT routing.  You might be stuck there.

On a server, this isn't just a Docker problem, but there are also mutlple levels of NAT usually going on in clusters, including network load balancers.

In Swarm, you can set a service to use host mode networking so it'll listen on the raw host IP:PORT, and that container will have access to the incoming 'client' IP, but if there's a LB in front of the Swarm (and there usually is) then you'll likely need to fallback to some other method, like asking the LB to [put the client IP in the HTTP header rather than TCP](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For).

Here's how the ports would look in the Swarm Service:
```
ports:
      - target: 80
        published: 80
        mode: host
      - target: 443
        published: 443
        mode: host
```

And here's an [example of me using it in Traefik 1.x](https://github.com/BretFisher/dogvscat/blob/master/stack-proxy-global.yml#L80-L88).
